staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -270,21 +279,192 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+					return pathOfFieldPath, allErrs
+				}
+			} else if tok != scanner.Ident {
+				allErrs = append(allErrs, nonexistentFieldError)
+				return pathOfFieldPath, allErrs
+			} else {
+				validFieldPath = validFieldPath.Child(fieldName)
+			}
+
+			if schema.Properties != nil {
+				propertySchema, ok := schema.Properties[fieldName]
+				if ok {
+					schema = &propertySchema
+				} else {
+					allErrs = append(allErrs, nonexistentFieldError)
+					return pathOfFieldPath, allErrs
+				}
+			} else if schema.AdditionalProperties != nil {
+				additionalProperty, ok := schema.AdditionalProperties.Structural.Properties[fieldName]
+				if ok {
+					schema = &additionalProperty
+				} else {
+					allErrs = append(allErrs, nonexistentFieldError)
+					return pathOfFieldPath, allErrs
+				}
+			} else {
+				allErrs = append(allErrs, nonexistentFieldError)
+				return pathOfFieldPath, allErrs
+			}
+
+			if isMapSyntax {
+				if tok == scanner.EOF {
+					allErrs = append(allErrs, nonexistentFieldError)
+					return pathOfFieldPath, allErrs
+				}
+				s.Scan()
+				if s.TokenText() != "]" {
+					allErrs = append(allErrs, nonexistentFieldError)
+					return pathOfFieldPath, allErrs
+				}
+			}
+		case "array":
+			isMapSyntax := false
+			if s.TokenText() == "[" {
+				isMapSyntax = true
+			} else if s.TokenText() != "." {
+				allErrs = append(allErrs, nonexistentFieldError)
+				return pathOfFieldPath, allErrs
+			}
+
+			tok = s.Scan()
+			if tok == scanner.EOF {
+				allErrs = append(allErrs, nonexistentFieldError)
+				return pathOfFieldPath, allErrs
+			}
+			fieldName := s.TokenText()
+			if isMapSyntax {
+				if tok == scanner.Char && len(fieldName) >= 2 && fieldName[0] == '\'' &&
+					fieldName[len(fieldName)-1] == '\'' {
+					fieldName = fieldName[1 : len(fieldName)-1]
+					validFieldPath = validFieldPath.Key(fieldName)
+				} else {
+					allErrs = append(allErrs, nonexistentFieldError)
+					return pathOfFieldPath, allErrs
+				}
+			} else if tok != scanner.Ident {
+				allErrs = append(allErrs, nonexistentFieldError)
+				return pathOfFieldPath, allErrs
+			} else {
+				validFieldPath = validFieldPath.Child(fieldName)
+			}
+			if schema.Items != nil {
+				schema = schema.Items
+			} else {
+				allErrs = append(allErrs, nonexistentFieldError)
+				return pathOfFieldPath, allErrs
+			}
+			if isMapSyntax {
+				if tok == scanner.EOF {
+					allErrs = append(allErrs, nonexistentFieldError)
+					return pathOfFieldPath, allErrs
+				}
+				s.Scan()
+				if s.TokenText() != "]" {
+					allErrs = append(allErrs, nonexistentFieldError)
+					return pathOfFieldPath, allErrs
+				}
+			}
+		default:
+			allErrs = append(allErrs, nonexistentFieldError)
+			return pathOfFieldPath, allErrs
+		}
+	}
+
+	if !found {
+		allErrs = append(allErrs, nonexistentFieldError)
+	}
+
+	return validFieldPath, allErrs
+}
+
+func fieldErrorForReason(fldPath *field.Path, value interface{}, detail string, reason *field.ErrorType) *field.Error {
```

Comment:
some of the reasons (TooLong, NotFound, Duplicate, NotSuported, TooMany) discard the message passed in, which is weird and probably undesirable... shouldn't we prioritize the message they provided over the canned message for these types?

Reply:
I think most of them are taking messages passed in, for those who don't(like required), the error is clear self explained. I will take another look after code freeze to see if a refactoring is needed.

Reply:
Have dropped support for MaxLength/TooMany/NotSupported/NotFound.

Reply:
For posterity: In an offline discussion we agreed to:

- Drop support for MaxLength/TooMany/NotSupported/NotFound. These are not a clear fit for this field.
- We will figure out how to pipe the message into Duplicate after 1.28 code freeze.

If we do both these things, it resolves all the discarded messages.